### PR TITLE
Mapping SQL Command errors to its specific model's error class

### PR DESCRIPTION
**Pull request by hlegius**
There are some minor improvements to do around this issue. They are:

 - [x] Change switch-case over class constant/fetch
 - [x] Move all error classes into `model/error.rb`
 - [x] Ensure the message detail level in each one of `Hanami::Model::Error` type
 - [x] Update README.md to reflect these changes.

Improve and close both #289 and #284

### CODE REVIEW

**Comment by joneslee85** [comment]
@hlegius looks good, can you please wrap it up :) :+1: ?

**Comment by hlegius** [comment]
@hanami/core-team 

About how descriptive our error messages are:

screen shot 2016-02-03 at 22 18 08

generated by:

```ruby
puts '------------------------------'
puts 'Type Violation Exception (UUID problem)'
puts '------------------------------'

begin
  model = ModelTests.find('non-uuid')
rescue => e
  puts "Class name: #{e.class}"
  puts "Error Message: #{e.message}"
end

puts ''
puts '------------------------------'
puts 'Not Null Violation Exception'
puts '------------------------------'

relation = ModelRelation.new
begin
  ModelRelations.create(relation)
rescue => e
  puts "Class name: #{e.class}"
  puts "Error Message: #{e.message}"
end

puts ''
puts '------------------------------'
puts 'Foreign Key Constraint Violation'
puts '------------------------------'

begin
  relation = ModelRelation.new(name: 'Foolano', model_test_id: '889352c4-ae20-3298-9fa3-5dace8573d18')
  ModelRelations.create(relation)
rescue => e
  puts "Class name: #{e.class}"
  puts "Error Message: #{e.message}"
end

puts ''
puts '------------------------------'
puts 'Invalid Query Error'
puts '------------------------------'

begin
  ModelRelations.invalid_query
rescue => e
  puts "Class name: #{e.class}"
  puts "Error Message: #{e.message}"
end
```

File: README.md [hunk 1]
```diff
@@ -470,6 +470,20 @@ If you need more information regarding those methods, you can use comments from
 
 Think of an adapter for Redis, it will probably employ different strategies to filter records than an SQL query object.
 
+### Model Error Coercions
+
+We encapsulate all adapter's errors into Hanami specialized error classes.
```
Comment (joneslee85) [inline README.md]: how about:

`All adapters' errors are encapsulated into Hanami error classes`

Comment (hlegius) [inline README.md]: :+1:

File: lib/hanami/model/adapters/sql/command.rb [hunk 2]
```diff
@@ -9,6 +9,13 @@ module Sql
         # @api private
         # @since 0.1.0
         class Command
+          ERROR_MAPPER = {
+            'Sequel::UniqueConstraintViolation'     => Hanami::Model::UniqueConstraintViolationError,
```
Comment (joneslee85) [inline lib/hanami/model/adapters/sql/command.rb]: can't we do class instead here?

Comment (hlegius) [inline lib/hanami/model/adapters/sql/command.rb]: I've got a _weirdo_ load issue when tried to keep class => class.

File: lib/hanami/model/adapters/sql/command.rb [hunk 3]
```diff
@@ -69,28 +76,7 @@ def delete
...
-            when Sequel::ForeignKeyConstraintViolation
-              Hanami::Model::ForeignKeyConstraintViolationError
-            when Sequel::NotNullConstraintViolation
-              Hanami::Model::NotNullConstraintViolationError
-            when Sequel::UniqueConstraintViolation
-              Hanami::Model::UniqueConstraintViolationError
-            else
-              Hanami::Model::InvalidCommandError
-            end.new(error.message)
+            raise (ERROR_MAPPER[e.class.name] || Hanami::Model::InvalidCommandError).new(e.message)
```
Comment (pascalbetz) [inline lib/hanami/model/adapters/sql/command.rb]: A little late to the party...

Can we use ```fetch``` with a default value?

```ruby
ERROR_MAPPER.fetch(e.class.name, Hanami::Model::InvalidCommand).new(e.message)
```

Comment (joneslee85) [inline lib/hanami/model/adapters/sql/command.rb]: @pascalbetz :+1: that's even BETTER!

can you please submit a PR? Thx

Comment (pascalbetz) [inline lib/hanami/model/adapters/sql/command.rb]: Will do.

**Pull request merged**